File: 209-minimum-size-subarray-sum/minimum-size-subarray-sum.cpp
class Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums) {
        int n = nums.size();
        int start = 0, end = 0;
        int minLen = INT_MAX;
        int currSum = 0;
        while (start < n) {
            while (end < n && currSum < target) {
                currSum += nums[end];
                end += 1;
            }
            if (currSum >= target)
                minLen = min(minLen, end - start);
            currSum -= nums[start];
            start += 1;
        }


        return minLen == INT_MAX ? 0 : minLen;
    }
};

File: 209-minimum-size-subarray-sum/minimum-size-subarray-sum.java
class Solution {
    public int minSubArrayLen(int target, int[] nums) {
        int n = nums.length;

        int start = 0, end = 0;
        int len = n + 1;
        int sum = 0;
        while (start < n) {
            while (end < n && sum < target) {
                sum += nums[end];
                end++;
            }

            if (sum >= target) {
                len = Math.min(len, end - start);
            }

            sum -= nums[start];
            start++;
        }

        return (len == n + 1) ? 0 : len;
    }
}